### PR TITLE
Optimize cache update. 

### PR DESCRIPTION
**Pull request by wang2yn84**
We used to insert cache inside attention, then use updated cache for calculation. With the help of flash attention/ragged attention, we can delay the cache insertion to the end of each step. By switching to left aligned stacked cache, we can minimize the data transfer to HBM and therefore improve performance. The decode step time reduced from 52ms to 42ms. The left aligned cache also improves the insert efficiency. The overall benchmark performance is boosted by 15%.

### CODE REVIEW

File: jetstream_pt/layers.py [hunk 1]
```diff
@@ -379,18 +395,36 @@ def repeat_kv(x: torch.Tensor, n_rep: int) -> torch.Tensor:
...
+        output_specs=(q_pspec, (q_pspec, q_pspec)),
+        q_shard_axis=self.q_shard_axis,
+        kv_shard_axis=self.kv_shard_axis,
+    )
+    self.ragged_attention_new = ak.RaggedAttentionKernel(
         env,
-        input_specs=(*([qkv_pspec] * 3), *([others_pspec] * 4)),
-        output_specs=(qkv_pspec, (others_pspec, others_pspec)),
-        sharding_axis=self.shard_axis,
+        input_specs=(q_pspec, q_pspec, q_pspec, *([others_pspec] * 7)),
```
Comment (FanhaiLu1) [inline jetstream_pt/layers.py:422]: correct me if I'm wrong, the ragged_attention_new doesn't support generate_cache_stacked

**Comment by FanhaiLu1** [comment]
> We used to insert cache inside attention, then use updated cache for calculation. With the help of flash attention/ragged attention, we can delay the cache insertion to the end of each step. By switching to left aligned stacked cache, we can minimize the data transfer to HBM and therefore improve performance. The decode step time reduced from 52ms to 42ms. The overall benchmark performance is boosted by 15%.



> we can delay the cache insertion to the end of each step

15% improvement is a great achievement! I assume the test side use stacked aligned + ragged attention, do you have any performance number with left aligned (without stacked) + ragged attention?

File: jetstream_pt/config.py [hunk 2]
```diff
@@ -86,10 +86,34 @@
 )
 flags.DEFINE_bool(
     "ring_buffer",
-    True,
+    False,
     "Whether to enable ring buffer",
     required=False,
 )
+flags.DEFINE_bool(
+    "flash_attention",
```
[the remark this answers is not in the document]
Comment (wang2yn84) [reply jetstream_pt/config.py:94]: No, ragged attention has better performance than flash attention. As I indicated in the description, it only takes effect at test mode. Which means user cannot directly enable it in either interactive, offline or server mode.

File: jetstream_pt/config.py [hunk 3]
```diff
@@ -86,10 +86,34 @@
...
+    True,
+    "Whether to enable flas attention. Only takes effect at test mode",
+    required=False,
+)
+flags.DEFINE_bool(
+    "generate_cache_stacked",
```
[the remark this answers is not in the document]
Comment (wang2yn84) [reply jetstream_pt/config.py:100]: It reduces the DMA transfer time. Minimize the number of DMA transfer helps.

File: jetstream_pt/layers.py (hunk 1 above)
Comment (wang2yn84) [reply jetstream_pt/layers.py:422]: ragged_attention_new is for new cache in the current step, which has length of 1, so there is nothing to stack.

File: jetstream_pt/attention_kernel.py [hunk 4]
```diff
@@ -105,60 +106,99 @@ def run():
...
     mask_value: float = DEFAULT_MASK_VALUE,
     normalize_var: bool = True,
+    testing: bool = False,
+    quantized: bool = False,
 ) -> tuple[jax.Array, tuple[jax.Array, jax.Array]]:
   """Ragged multi query attention."""
   with jax.named_scope("ragged_mqa"):
-    batch_size, num_heads, head_dim = q.shape
-    seq_len = k.shape[1]
+    batch_size, time, head_dim = q.shape
```
[the remark this answers is not in the document]
Comment (wang2yn84) [reply jetstream_pt/attention_kernel.py:137]: After vmap, the number of heads dimension are gone. So it's indeed the sequence length dimension, which we can also call it "time".

File: jetstream_pt/attention_kernel.py [hunk 5]
```diff
@@ -105,60 +106,99 @@ def run():
...
+    seq_len = k.shape[-2]
+
+    stacked = False
+    if k.ndim == 5:
```
[the remark this answers is not in the document]
Comment (wang2yn84) [reply jetstream_pt/attention_kernel.py:141]: I'm not sure why the block the quantization matters. If the cache is stacked, it will have layer, batch, number of heads, time, head dim these 5 dimensions no matter if it's quantized or not.

File: jetstream_pt/attention_kernel.py [hunk 6]
```diff
@@ -191,33 +225,241 @@ def scaler_index_map(b, i, *_):
...
+    v_scaler_ref,
+    o_ref,
+    m_ref,
+    l_ref,
+    bk: int,
+    mask_value: float,
+    normalize_var: bool,
+    quantized: bool,
+):
+  """Pallas kernel for flash attention."""
```
[the remark this answers is not in the document]
Comment (wang2yn84) [reply jetstream_pt/attention_kernel.py]: Thanks, updated!

File: jetstream_pt/attention_kernel.py [hunk 7]
```diff
@@ -191,33 +225,241 @@ def scaler_index_map(b, i, *_):
...
+
+  # length = lengths_ref[b]
+  # Always start from 0, left aligned
+  length = end_ref[b]
+
+  @pl.when(i * bk < length)
+  def run():
+    q = q_ref[...].astype(jnp.float32)
+    k = k_ref[...].astype(jnp.float32)
+    v = v_ref[...].astype(jnp.float32)
```
[the remark this answers is not in the document]
Comment (wang2yn84) [reply jetstream_pt/attention_kernel.py:283]: All the arithmetic operation only supports f32 and it reports error if force to be bf16. Confirmed with XLA team about the constraint: b/340263269 and b/341729764.

File: jetstream_pt/attention_kernel.py [hunk 8]
```diff
@@ -191,33 +225,241 @@ def scaler_index_map(b, i, *_):
...
+    )
+    return b_next, i_next
+
+  def kv_index_map(b, i, layer_ref, start_ref, end_ref, *_):
+    b_next, i_next = _compute_ragged_block_indices(b, i, end_ref)
+    if stacked:
+      return layer_ref[0], b_next, i_next, 0
+    return b_next, i_next, 0
+
+  def kv_scale_index_map(b, i, layer_ref, start_ref, end_ref, *_):
```
[the remark this answers is not in the document]
Comment (wang2yn84) [reply jetstream_pt/attention_kernel.py:382]: i_next doesn't get different value. It's in different position because the scale has the shape of batch, 1, kv_length. And the grid[1] applied to the last dimension here. That's why we give the i_next in this dimension.

File: jetstream_pt/attention_kernel.py [hunk 9]
```diff
@@ -191,33 +225,241 @@ def scaler_index_map(b, i, *_):
...
+      pl.BlockSpec(kv_index_map, kv_bp),  # v
+      pl.BlockSpec(kv_scale_index_map, ks_bp),  # k_scaler
+      pl.BlockSpec(kv_scale_index_map, ks_bp),  # v_scaler
+  ]
+
+  inputs = (
+      jnp.array([layer]),
+      start,
+      end,
+      end,  # line_end, not actually used
```
[the remark this answers is not in the document]
Comment (wang2yn84) [reply jetstream_pt/attention_kernel.py:407]: Pallas kernel doesn't take None.

File: jetstream_pt/attention_kernel.py [hunk 10]
```diff
@@ -310,15 +586,78 @@ def dense_attention(xq, keys, values, k_scaler=None, v_scaler=None, mask=None):
   return output
 
 
+def flash_attention(
```
[the remark this answers is not in the document]
Comment (wang2yn84) [reply jetstream_pt/attention_kernel.py:589]: Flash attention has the capability of blockwise compute the local softmax, which is exactly what we are doing here. In terms of how to divide the block, it's up to the user. We leveraged this to divide the attention calculation to existing cache and new cache. So this is indeed the flash attention.

File: jetstream_pt/cache_manager.py [hunk 11]
```diff
@@ -89,48 +96,190 @@ def __init__(
...
+        self.cache_k._elem, self.cache_v._elem = torch_xla2.interop.call_jax(
+            self.update_single_cache_line,
+            self.cache_k._elem,
+            self.cache_v._elem,
+            self.new_ks._elem,
+            self.new_vs._elem,
+            self.input_pos,
+        )
+
+  def update(self, key, value, layer_id: int):
```
[the remark this answers is not in the document]
Comment (wang2yn84) [reply jetstream_pt/cache_manager.py:213]: I was thinking about merging the Int8KVCacheGenerate and KVCacheGenerate cuz there are a lot of shared code. I can combine all 4 additional flags (lazy_cache_update, generate_cache_stacked, new_cache_stacked, flash_attention) into 1 to simplify the logic, cuz these flags only helps for my experimentation. It should not be exposed to user. Wdyt?

**Comment by wang2yn84** [comment]
> > We used to insert cache inside attention, then use updated cache for calculation. With the help of flash attention/ragged attention, we can delay the cache insertion to the end of each step. By switching to left aligned stacked cache, we can minimize the data transfer to HBM and therefore improve performance. The decode step time reduced from 52ms to 42ms. The overall benchmark performance is boosted by 15%.
> 
> > we can delay the cache insertion to the end of each step
> 
> 15% improvement is a great achievement! I assume the test side use stacked aligned + ragged attention, do you have any performance number with left aligned (without stacked) + ragged attention?



> > We used to insert cache inside attention, then use updated cache for calculation. With the help of flash attention/ragged attention, we can delay the cache insertion to the end of each step. By switching to left aligned stacked cache, we can minimize the data transfer to HBM and therefore improve performance. The decode step time reduced from 52ms to 42ms. The overall benchmark performance is boosted by 15%.
> 
> > we can delay the cache insertion to the end of each step
> 
> 15% improvement is a great achievement! I assume the test side use stacked aligned + ragged attention, do you have any performance number with left aligned (without stacked) + ragged attention?

When cache is left aligned + unstacked, the data transfer overhead is non neglegible. I tried flash attention, which is 90ms for each step. These overhead has nothing to do with which attention you are using.

File: jetstream_pt/config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wang2yn84) [reply jetstream_pt/config.py:100]: Also the XLA handles cache insertion for all the layers much more efficiently than iterating over layer dimension by user.

File: jetstream_pt/cache_manager.py [hunk 12]
```diff
@@ -205,23 +499,133 @@ def empty(cls, shape, device, bf16_enable, env):
...
+              self.cache_v._elem,
+              self.new_ks._elem,
+              self.new_vs._elem,
+              self.k_scaler._elem,
+              self.v_scaler._elem,
+              self.new_k_scaler,
+              self.new_v_scaler,
+          ]
+          (
+              self.cache_k._elem,
```
Comment (qihqi) [inline jetstream_pt/cache_manager.py:578]: `._elem` seems spurious; as `_elem` is already a jax array.

So it's either: `x._elem = foo(jax_array_inputs)` OR `x = call_jax(foo, torch_tensor_inputs)`

File: jetstream_pt/layers.py [hunk 13]
```diff
@@ -367,9 +367,25 @@ def apply_rotary_emb(
 def repeat_kv(x: torch.Tensor, n_rep: int) -> torch.Tensor:
   """torch.repeat_interleave(x, dim=2, repeats=n_rep)."""
 
-  bs, n_kv_heads, slen, head_dim = x.shape
+  bs, n_kv_heads, slen, head_dim = (
```
Comment (qihqi) [inline jetstream_pt/layers.py:370]: bs, n_kv_heads, slen, head_dim, *_ = x.shape

File: jetstream_pt/layers.py [hunk 14]
```diff
@@ -367,9 +367,25 @@ def apply_rotary_emb(
...
+      x.shape[-4],
+      x.shape[-3],
+      x.shape[-2],
+      x.shape[-1],
+  )
+  if x.ndim == 5:
```
Comment (qihqi) [inline jetstream_pt/layers.py:376]: stacked = x.ndim == 5

File: jetstream_pt/layers.py [hunk 15]
```diff
@@ -367,9 +367,25 @@ def apply_rotary_emb(
...
+    stacked = True
+  else:
+    stacked = False
   if n_rep == 1:
     return x
+  if stacked:
```
Comment (qihqi) [inline jetstream_pt/layers.py:382]: or just put the ndim == 5 here and remove the stacked var

File: jetstream_pt/layers.py (hunk 13 above)
Comment (wang2yn84) [reply jetstream_pt/layers.py:370]: I see, should be *_, bs, n_kv_heads, slen, head_dim.= x.shape ?

File: jetstream_pt/cache_manager.py (hunk 12 above)
Comment (wang2yn84) [reply jetstream_pt/cache_manager.py:578]: Make sense, decided to remove _elem since it's violating lint anyway.

File: jetstream_pt/layers.py (hunk 14 above)
Comment (wang2yn84) [reply jetstream_pt/layers.py:376]: Better! Thanks!

File: jetstream_pt/layers.py (hunk 15 above)
Comment (wang2yn84) [reply jetstream_pt/layers.py:382]: I'd probably prefer to keep stacked to make the code more clear.

**Comment by wang2yn84** [comment]
Fixed based on your comments, all the unit tests and lint errors. Please let me know if you have any other comment/suggestions. @qihqi @FanhaiLu1

**Review by qihqi — APPROVED** [review]

**Comment by qihqi** [comment]
there is some updates on deps/Jetstream is that intentional?

File: jetstream_pt/cache_manager.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (FanhaiLu1) [reply jetstream_pt/cache_manager.py:226]: Thanks. My main concerns is current code logic is too complex for read and maintain. The cache manager is very straightforward implementation before, but right now the logic is very complex. Let's only keep the most optimized code in the repo.

**Comment by FanhaiLu1** [comment]
> Fixed based on your comments, all the unit tests and lint errors. Please let me know if you have any other comment/suggestions. @qihqi @FanhaiLu1

There are new lint error, can you fix it?

File: jetstream_pt/attention_kernel.py [hunk 16]
```diff
@@ -191,33 +225,241 @@ def scaler_index_map(b, i, *_):
...
+    testing: bool = False,
+    quantized: bool = False,
+) -> tuple[jax.Array, tuple[jax.Array, jax.Array]]:
+  """Ragged multi query attention."""
+  batch_size, time, head_dim = q.shape
+  # assert end.shape == (batch_size,)
+  seq_len = k.shape[-2]
+
+  stacked = False
+  if k.ndim == 4:
```
[the remark this answers is not in the document]
Comment (wang2yn84) [reply jetstream_pt/attention_kernel.py:355]: It's not a must. By reducing the number of heads dimension to 1 the MHA becomes MQA. That's just for compatibility.

**Comment by wang2yn84** [comment]
> > Fixed based on your comments, all the unit tests and lint errors. Please let me know if you have any other comment/suggestions. @qihqi @FanhaiLu1
> 
> There are new lint error, can you fix it?

Fixed all the lint issues.

**Comment by wang2yn84** [comment]
I will remove precompute_ragged_block_indices, clear up the ragged attention impl (e.g. remove the one for the ring buffer) and simplify the flags for non ring buffer case therefore simplify the cache manager in the subsequent PR. Will push this PR first since it's been standing alone for a while.
